### ISSUE TITLE
Interview questions

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#5584]
Interview questions<br>
<br>
https://ift.tt/2kUDrMr<br>
<br>
P&#275;teris &#325;ikiforovs<br>
<br>
<div>
<div>
<p>When I read <a href="https://www.reddit.com/r/cscareerquestions/">/r/cscareerquestions</a>, it gives me the impression that all companies in Silicon Valley only ask algorithmic questions at job interviews.</p>
<p>My experience as someone living in Europe is that you are generally given a project to complete at home or on-site. Then they take a look at your code and decide whether to give you an offer or not.</p>
<p>Here are the questions and tasks I've been given at job interviews in my career.</p>
<p>If there's a Skype interview before the on-site interview, it means I interviewed for a company in another country.</p>
<p>Webdev shop (2012)</p>
<p>This was my first ever proper job interview where I sent in my CV, dutifully did the take-home project and got an offer. It was after I returned from studying abroad and was looking for a part time job while finishing my Bachelor's degree.</p>
<p>I was offered a part-time job but I declined it because I went to work for another company, the next one on this list.</p>
<p>Review HTML/CSS/JavaScript code</p>
<p>The first task was to take a look at an order form written in HTML/CSS/JavaScript and point out everything that I see looks wrong and offer my suggestions to fix it.</p>
<p>Here is a snippet:</p>
<pre><code><span>&lt;!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"&gt;</span> <span>&lt;<span>html</span> <span>xmlns</span>=<span>"http://www.w3.org/1999/xhtml"</span> <span>xml:lang</span>=<span>"en"</span> <span>lang</span>=<span>"en"</span>&gt;</span> <span>&lt;<span>head</span>&gt;</span> <span>&lt;<span>title</span>&gt;</span>EDU<span>&lt;/<span>title</span>&gt;</span> <span>&lt;<span>meta</span> <span>http-equiv</span>=<span>"content-type"</span> <span>content</span>=<span>"application/xhtml+xml; charset=utf-8"</span> /&gt;</span> <span>&lt;<span>link</span> <span>rel</span>=<span>"stylesheet"</span> <span>href</span>=<span>"base.css"</span> <span>type</span>=<span>"text/css"</span> /&gt;</span> <span>&lt;<span>script</span> <span>type</span>=<span>"text/javascript"</span> <span>src</span>=<span>"jquery.js"</span>&gt;</span><span>&lt;/<span>script</span>&gt;</span> <span>&lt;<span>script</span> <span>type</span>=<span>"text/javascript"</span> <span>src</span>=<span>"base.js"</span>&gt;</span><span>&lt;/<span>script</span>&gt;</span> <span>&lt;/<span>head</span>&gt;</span> <span>&lt;<span>body</span>&gt;</span> <span>&lt;<span>div</span> <span>id</span>=<span>"page-wrapper"</span> <span>style</span>=<span>"border:1px solid silver;"</span>&gt;</span> <span>&lt;<span>h1</span>&gt;</span>Pas&#363;t&#299;juma forma<span>&lt;/<span>h1</span>&gt;</span> <span>&lt;<span>div</span> <span>id</span>=<span>"errors"</span> <span>style</span>=<span>"color:#EB8282;"</span>&gt;</span><span>&lt;/<span>div</span>&gt;</span> <span>&lt;<span>form</span> <span>action</span>=<span>"index.html"</span> <span>method</span>=<span>"post"</span>&gt;</span> <span>&lt;<span>div</span> <span>class</span>=<span>"row"</span>&gt;</span> <span>&lt;<span>label</span>&gt;</span>V&#257;rds, Uzv&#257;rds: <span>&lt;<span>font</span> <span>color</span>=<span>"#F35151"</span>&gt;</span>*<span>&lt;/<span>font</span>&gt;</span><span>&lt;/<span>label</span>&gt;</span> <span>&lt;<span>input</span> <span>type</span>=<span>"text"</span> <span>name</span>=<span>"full-name"</span> <span>value</span>=<span>""</span> <span>id</span>=<span>"full-name"</span> /&gt;</span> <span>&lt;<span>div</span> <span>class</span>=<span>"clearfloat"</span>&gt;</span><span>&lt;/<span>div</span>&gt;</span> <span>&lt;/<span>div</span>&gt;</span> ... </code></pre>
<p>I pointed out that, you should have <code>&lt;meta charset="utf-8"&gt;</code> before <code>&lt;title&gt;</code>, the official JavaScript MIME type is <code>application/javascript</code> not <code>text/javascript</code>, you should not use inline styles and the <code>&lt;font&gt;</code> tag, etc.</p>
<p>Registration + login in PHP</p>
<p>The other task was to create a basic CRUD app in PHP that lets users sign up and log in. You should check whether a username is taken with AJAX. After registration, a confirmation email should be sent. When you're logged in, you should see the latest news taken from a remote RSS feed and parsed using PHP DOM libraries. The HTML should also be generated with PHP DOM.</p>
<p>I didn't use any frameworks or libraries, but followed basic MVC principles. I think I named the router <code>FrontController</code>. I also included CSRF and rate limiting.</p>
<p>Language technology company (2012)</p>
<p>This is a language technology company that builds natural language processing and machine translation tools for the Baltic languages.</p>
<p>I had two hours to solve the three tasks below on-site.</p>
<p>Latvian inflector</p>
<p>Latvian is a highly inflected language, meaning that words have many different word endings. I had to create a Latvian noun and adjective inflector in C# using OOP princples.</p>
<p>Luckily, I know Latvian grammar by heart and as a native speaker I recreate the following noun declension table easily:</p>
<table>
<thead>
<tr>
<th>Declension</th>
<th>I</th>
<th>II</th>
<th>III</th>
<th>IV</th>
<th>V</th>
<th>VI</th>
</tr>
</thead>
<tbody>
<tr>
<td>Nominative</td>
<td>-s</td>
<td>-is</td>
<td>-us</td>
<td>-a</td>
<td>-e</td>
<td>-s</td>
</tr>
<tr>
<td>Genitive</td>
<td>-a</td>
<td>-a</td>
<td>-us</td>
<td>-as</td>
<td>-es</td>
<td>-s</td>
</tr>
<tr>
<td>Dative</td>
<td>-am</td>
<td>-im</td>
<td>-um</td>
<td>-ai</td>
<td>-ei</td>
<td>-ij</td>
</tr>
<tr>
<td>Accusative</td>
<td>-u</td>
<td>-i</td>
<td>-u</td>
<td>-u</td>
<td>-i</td>
<td>-i</td>
</tr>
<tr>
<td>Locative</td>
<td>-&#257;</td>
<td>-&#299;</td>
<td>-&#363;</td>
<td>-&#257;</td>
<td>-&#275;</td>
<td>-&#299;</td>
</tr>
</tbody>
</table>
<table>
<thead>
<tr>
<th>Declension</th>
<th>I</th>
<th>II</th>
<th>III</th>
<th>IV</th>
<th>V</th>
<th>VI</th>
</tr>
</thead>
<tbody>
<tr>
<td>Nominative</td>
<td>-i</td>
<td>-i</td>
<td>-i</td>
<td>-as</td>
<td>-es</td>
<td>-is</td>
</tr>
<tr>
<td>Genitive</td>
<td>-u</td>
<td>-u</td>
<td>-u</td>
<td>-u</td>
<td>-u</td>
<td>-u</td>
</tr>
<tr>
<td>Dative</td>
<td>-iem</td>
<td>-iem</td>
<td>-iem</td>
<td>-&#257;m</td>
<td>-&#275;m</td>
<td>-&#299;m</td>
</tr>
<tr>
<td>Accusative</td>
<td>-us</td>
<td>-us</td>
<td>-us</td>
<td>-as</td>
<td>-es</td>
<td>-is</td>
</tr>
<tr>
<td>Locative</td>
<td>-os</td>
<td>-os</td>
<td>-os</td>
<td>-&#257;s</td>
<td>-&#275;s</td>
<td>-&#299;s</td>
</tr>
</tbody>
</table>
<p>It's actually a little more complicated than that but I basically had to turn this table into OOP code.</p>
<p>Orientation of a polygon</p>
<p>This was an algorithmic question.</p>
<p>I had a list of points and I had to print whether the orientation of the polygon that was made up of these points was clockwise or counter-clockwise.</p>
<p>I think I couldn't figure out the solution.</p>
<p>Word problem</p>
<p>Two people had pens, pencils and erasers and I had to figure out how many erasers the second person had, I think. It was a simple middle school math problem that can be solved with a simple equation or just brute forced.</p>
<p>Apparently, there was no catch but still many people can't solve it.</p>
<p>Local directory service (2014)</p>
<p>When an ex-boss sends you a message on Skype (that you never use), what's the first thing that comes to your mind? In my case, I immediately started worrying that my code that I wrote three years ago no longer works.</p>
<p>But it turns out he was now at a different company and was looking for developers.</p>
<p>The take home task was supposed to be a simple MVC CRUD app in PHP.</p>
<p>I declined to do the task because they offered to pay under the table.</p>
<p>Charts library startup (2015)</p>
<p>I didn't apply for this job but I wanted to see if I can solve this interview problem.</p>
<p>You need to create an HTML page with a 20x20 grid of rotating doughnut charts.</p>
<ul>
<li>Doughnuts have different colors</li>
<li>Rotation speed is 1 full rotation every 2 seconds</li>
<li>When you click on it, the color and rotation direction is changed</li>
<li>There's a static image in the background</li>
</ul>
<p><img src="https://peteris.rocks/blog/interview-questions/pies.png"></p>
<p>I had to read up on <code>requestAnimationFrame</code> otherwise my CPU fan got way too loud.</p>
<p><a href="http://rawgit.com/pdonald/b57f56f5c9113847b085/raw/69e4e43d8fd2d7347df6101ca49b99b7d76c16d4/donuts.html">Here's my solution</a>.</p>
<p>Language technology company (2015)</p>
<p>I didn't apply for this job but I tried my hand at this interview problem.</p>
<p>It's a machine learning classification task. You have 16 numeric features (named <code>F1</code> to <code>F16</code>) and a binary label.</p>
<p>Your task is to</p>
<ul>
<li>analyze &amp; find redundant features</li>
<li>train a classifier</li>
<li>write the code that does predictions on the test set yourself</li>
</ul>
<p>The dataset was generated in Excel (random numbers, some were linear combinations) and you could easily get a 100% accuracy with most algorithms.</p>
<p>Large tech company (2015)</p>
<p>First Skype interview consisted of various questions</p>
<ul>
<li>What happens when you type <code>domain.com</code> in the browser (in as much detail as you can)</li>
<li>How <code>list</code> in Python is implemented? What happens when a new item is added?</li>
<li>Explain prototipal inheritance in JavaScript</li>
<li>Your eshop can no longer contact the payment gateway but your users can. How can you make make payments work?</li>
<li>...some more. I don't remember them anymore.</li>
</ul>
<p>This was insanely fun. First, you'd need to look up <code>domain.com</code> using DNS. I had prepared for this so I talked about the whole path down to the network layer. Not mentioning ARP and routing tables would be a red flag. Then you'd need to establish a connection over TCP but which port? It's usually <code>80</code> but it could also be <code>443</code>. I talked about the HSTS header, HSTS preload list and certificate pinning, etc.</p>
<p>Second Skype interview was a coding interview</p>
<ul>
<li>Count words or something, not too trivial but not complex either</li>
<li>Implement a hash table</li>
<li>Implement an algorithm to find Fibonacci numbers, also without recursion</li>
</ul>
<p>I was told I passed both interviews and the next step was supposted to be an on-site interview.</p>
<p>But my emails about the next steps went unanwered even when I tried a year later.</p>
<p>Startup (2016)</p>
<p>I applied for what I guess was mostly a front-end developer position.</p>
<p>I had two quick coding questions in the Skype interview.</p>
<ul>
<li>How would you find most frequent words in a text file? Can you do it with Linux tools?</li>
<li>Create two SQL tables (products and orders). What indexes would you add? How to make a query with <code>SUM()</code> fast?</li>
</ul>
<p>I was then invited to an on-site interview.</p>
<p>The task on the interview day was to create a small application that has coordinates x, y in a table on the left and a graph with these points on the right. When you click on the graph, a new point is added. Or you can add it to the table.</p>
<p>Bonus features (pick two)</p>
<ul>
<li>create documentation</li>
<li>write unit tests</li>
<li>make it look nice with CSS</li>
<li>realtime with WebSockets</li>
</ul>
<p>I chose realtime and unit tests.</p>
<p>Fintech startup (2016)</p>
<p>This is a fintech startup that tries to classify transactions on your credit card statement.</p>
<p>They had a challenge on their website where they give you a small sample of their data and you need to create a machine leraning solution to classify the transactions in five categories (food, restaurants, transportation, utilities and other).</p>
<p>Fun, right?</p>
<p>Except the training data was not labeled. They wanted you to label all 6000 transactions yourself. What if a transaction is ambiguous: is it food or restaurant? You can clarify that with them by email.</p>
<p>They would compare your results with their hand crafted algorithm that's got &gt;95% accuracy.</p>
<p>The prize was &euro;200 and possibly a job opportunity.</p>
<p>When communicating with them, I got the impression that they did not know what they were doing. I learned later that I was completely right.</p>
<p>Monea (2016)</p>
<p>If you want interviewees to badmouth your company every chance they get, do what this company did.</p>
<ul>
<li>Have a recruiter that literally can't spell contact you</li>
<li>Have another recruiter contact you because the previous one was banned from LinkedIn</li>
<li>Make you wait for a month</li>
<li>Ask if you mind being yelled at, but the next day there's going to be no hard feelings</li>
<li>Give you a take-home project</li>
<li>Ignore you afterwards, not even a rejection email after you complete the project</li>
</ul>
<p>Swedbank (2016)</p>
<p>There was an encrypted message in the job posting and you had to decrypt it (it was Base64) and follow the instructions. I wrote about it on my blog <a href="https://peteris.rocks/blog/swedbank-ethical-hacker-puzzle/">here</a>.</p>
<p>I had a non-technical Skype interview with Swedbank but the interviewers looked bored and I wasn't really gunning for a penetration testing position so I didn't get an offer.</p>
<p>Social network startup (2017)</p>
<p>There were three Skype interviews.</p>
<p>Missing numbers</p>
<p>You have an array of numbers from 1 to <code>n</code>. One number is missing. Find which one.</p>
<ul>
<li>Follow up: can you do it without sorting?</li>
<li>Follow up: what problems could you face if the numbers were large?</li>
</ul>
<p>I solved it because I'd practiced this problem and seen others' solutions on Leetcode.</p>
<p>Calculator</p>
<p>We want you to create a simple calculator that can add, subtract, multiply numbers. Keep in mind that we may want to add abitrary new operators/functions in the future. You can have the input in any format that you want.</p>
<p>I started to talk about the postfix notation and the shunting-yard algorithm (whose name or implementation I did not know) but they started to stress the importance of adding new functions.</p>
<p>Then it occurred to me that I can write a simple Lisp interpreter.</p>
<p>I could solve this because I'd tinkered with writing insignificant toy parsers before and I'd read <a href="http://norvig.com/lispy.html">Peter Norvig's article</a>.</p>
<p>Webdev questions</p>
<p>This was another Skype interview, this time with the CTO.</p>
<ul>
<li>Our clients want to add arbitrary fields to user profiles. How'd you implement that?</li>
<li>Have you ever created RESTful apps?</li>
<li>Some questions about authentication, cookies, security, etc.</li>
<li>...some other similar questions.</li>
</ul>
<p>Feature</p>
<p>At the on-site interview, I had to add a small feature to their codebase.</p>
<p>In their product, they have a chat very much like Facebook Messenger. It's basically a copy of it.</p>
<p>They wanted me to add statuses to each chat message (sending, error, received, read). Just like on Messenger.</p>
<p>I implemented a hackish (which I was encouraged to do) solution that did not work. My problem was that the feedback loop was very slow (it'd take a minute for me to make a change in code and see it reflected in the browser) and I started on the wrong path and did not have time to finish with a clean solution.</p>
<p>I didn't think I'd get an offer but I actually did. I declined because I went to work for another company.</p>
<p>Summary</p>
<ul>
<li>Jobs interviewed for: 7</li>
<li>Offers: 4</li>
<li>Rejections or no answer: 3</li>
<li>Strike rate: 57%</li>
</ul>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via peteris.rocks<br>
<br>
November 1, 2019 at 06:06PM
